File: tests/contrato.py
from dataclasses import FrozenInstanceError
import sys
from colecoes import Catalogo, Medicao
from historico import Historico
from identidade import IdSensor


def exigir(condicao, mensagem):
    if not condicao:
        raise AssertionError(mensagem)


def base():
    a, b, diferente = IdSensor("LT-101"), IdSensor("LT-101"), IdSensor("LT-102")
    alias = a
    exigir(a is not b and alias is a, "00: distinguir identidade e alias")
    exigir(a == b and b == a and a == a and a != diferente, "00: igualdade pela tag")
    exigir(a != IdSensor("lt-101"), "00: preservar maiusculas")
    exigir(hash(a) == hash(b), "00: iguais exigem hashes iguais")
    exigir(len({a, b, diferente}) == 2, "00: conjunto deve reunir dois valores distintos")
    try:
        IdSensor("")
    except ValueError:
        pass
    else:
        raise AssertionError("00: recusar tag vazia")
    try:
        a.valor = "outra"
    except FrozenInstanceError:
        pass
    else:
        raise AssertionError("00: chave deve impedir reatribuicao usual")
    h = Historico()
    exigir(h.quantidade() == 0, "00: historico nasce vazio")
    h.registrar(Medicao(12, "%")); h.registrar(Medicao(12, "%"))
    exigir(h.quantidade() == 2, "00: preservar ocorrencias repetidas")
    c = Catalogo[Medicao]()
    exigir(c.quantidade() == 0 and c.ids() == set(), "00: catalogo nasce vazio")


def catalogo():
    c = Catalogo[Medicao]()
    exigir(c.buscar(IdSensor("ausente")) is None and c.quantidade() == 0,
           "01: buscar no vazio deve retornar None sem criar entrada")
    exigir(c.inserir(IdSensor("LT-101"), Medicao(12, "%")) is True,
           "01: primeira insercao deve retornar True; complete inserir")
    exigir(c.inserir(IdSensor("LT-101"), Medicao(99, "%")) is False,
           "01: recusar outra instancia da mesma tag")
    m = c.buscar(IdSensor("LT-101"))
    exigir(m is not None, "01: buscar pela tag equivalente deve encontrar; complete buscar")
    exigir(m == Medicao(12, "%"), "01: duplicata deve preservar valor E unidade originais")
    exigir(c.inserir(IdSensor("LT-101"), Medicao(800, "cm")) is False, "01: mudar unidade nao cria nova chave")
    exigir(c.buscar(IdSensor("LT-101")).unidade == "%", "01: unidade original preservada")
    exigir(c.inserir(IdSensor("LT-102"), Medicao(0, "%")) is True, "01: aceitar leitura zero")
    exigir(c.inserir(IdSensor("lt-101"), Medicao(7, "%")) is True, "01: tag minuscula e distinta")
    exigir(c.quantidade() == 3 and len(c.ids()) == 3, "01: contar chaves, nao tentativas")
    exigir(c.buscar(IdSensor("LT-102")).valor == 0, "01: zero nao representa ausencia")
    exigir(c.buscar(IdSensor("outra")) is None and c.quantidade() == 3, "01: busca ausente nao altera quantidade")
    exigir(IdSensor("LT-101") in c.ids(), "01: conjunto de ids acompanha o mapa")
    exigir(Catalogo().quantidade() == 0, "01: catalogos nao compartilham armazenamento")
    nomes = Catalogo[str]()
    exigir(nomes.inserir(IdSensor("X"), "Bancada A") is True, "01: genericidade deve aceitar texto")
    exigir(nomes.buscar(IdSensor("X")) == "Bancada A", "01: busca generica preserva texto")


def ultimas(h, n, valores):
    dados = h.ultimas(n)
    exigir(isinstance(dados, list), "02: ultimas deve devolver uma lista independente")
    exigir(dados == [Medicao(v, "%") for v in valores],
           f"02: ultimas({n}) deve retornar valores {valores}, na ordem de chegada, com unidade %")


def extensao():
    c = Catalogo[Medicao]()
    exigir(c.remover(IdSensor("vazio")) is False, "02: remover no vazio retorna False")
    c.inserir(IdSensor("A"), Medicao(12, "%")); c.inserir(IdSensor("B"), Medicao(20, "%"))
    exigir(c.remover(IdSensor("X")) is False and c.quantidade() == 2, "02: ausente preserva entradas")
    exigir(c.remover(IdSensor("A")) is True, "02: remover existente deve retornar True; complete remover")
    exigir(c.quantidade() == 1 and c.buscar(IdSensor("A")) is None, "02: atualizar quantidade e busca")
    exigir(c.ids() == {IdSensor("B")}, "02: ids deve refletir remocao")
    exigir(c.buscar(IdSensor("B")) == Medicao(20, "%"), "02: preservar outra chave")
    exigir(c.remover(IdSensor("A")) is False, "02: segunda remocao retorna False")
    exigir(c.inserir(IdSensor("A"), Medicao(33, "%")) is True, "02: permitir reinsercao")
    exigir(c.buscar(IdSensor("A")).valor == 33, "02: reinsercao guarda novo valor")
    h = Historico()
    ultimas(h, 0, []); ultimas(h, 4, [])
    for valor in (12, 12, 15):
        h.registrar(Medicao(valor, "%"))
    for n, esperados in [(0, []), (1, [15]), (2, [12, 15]), (3, [12, 12, 15]), (100, [12, 12, 15])]:
        ultimas(h, n, esperados)
    exigir(h.quantidade() == 3, "02: consultar nao descarta historico")
    copia = h.ultimas(3)
    copia[0] = Medicao(999, "%"); copia.clear()
    ultimas(h, 3, [12, 12, 15])
    externa = Medicao(40, "%"); h.registrar(externa)
    exigir(h.ultimas(1)[0] is externa, "02: Python guarda referencias aos registros imutaveis")
    try:
        externa.valor = 99
    except FrozenInstanceError:
        pass
    else:
        raise AssertionError("02: Medicao deve manter campos imutaveis")
    exigir(c.remover(IdSensor("A")) is True and h.quantidade() == 4,
           "02: remover cadastro nao apaga historico independente")
    h.registrar(Medicao(5, "%")); ultimas(h, 2, [40, 5])
    try:
        h.ultimas(-1)
    except ValueError:
        pass
    else:
        raise AssertionError("02: preservar rejeicao fornecida para limite negativo")


def main():
    etapa = sys.argv[1]
    exigir(etapa in ("00", "01", "02"), "etapa invalida")
    base()
    if etapa >= "01":
        catalogo()
    if etapa == "02":
        extensao()
    print(f"OK Python etapa {etapa}")


if __name__ == "__main__":
    try:
        main()
    except Exception as erro:
        print(f"FALHA Python: {erro}", file=sys.stderr)
        raise SystemExit(1)

File: tests/contrato.cpp
#include "colecoes.hpp"
#include "historico.hpp"
#include <iostream>
#include <stdexcept>
#include <unordered_set>
#include <vector>

void exigir(bool condicao, const std::string& mensagem) {
    if (!condicao) throw std::runtime_error(mensagem);
}
void base() {
    IdSensor a{"LT-101"}, b{"LT-101"}, diferente{"LT-102"}, minuscula{"lt-101"};
    const IdSensor& alias = a;
    exigir(&a != &b && &a == &alias, "00: distinguir identidade e alias");
    exigir(a == b && b == a && a == a && !(a == diferente), "00: igualdade pela tag");
    exigir(!(a == minuscula), "00: preservar maiusculas na tag exata");
    exigir(!(a < b) && !(b < a), "00: chaves iguais devem ser equivalentes no mapa");
    exigir(HashId{}(a) == HashId{}(b), "00: iguais exigem hashes iguais");
    std::unordered_set<IdSensor, HashId> unicos{a, b, diferente};
    exigir(unicos.size() == 2, "00: conjunto deve reunir dois valores distintos");
    bool rejeitou = false;
    try { IdSensor vazio{""}; } catch (const std::invalid_argument&) { rejeitou = true; }
    exigir(rejeitou, "00: recusar tag vazia");
    Historico h;
    exigir(h.quantidade() == 0, "00: historico nasce vazio");
    h.registrar({12, "%"}); h.registrar({12, "%"});
    exigir(h.quantidade() == 2, "00: historico deve guardar ocorrencias repetidas");
    Catalogo<Medicao> c;
    exigir(c.quantidade() == 0 && c.ids().empty(), "00: catalogo nasce vazio");
}
void catalogo() {
    Catalogo<Medicao> c;
    exigir(c.buscar(IdSensor{"ausente"}) == nullptr && c.quantidade() == 0,
           "01: buscar no vazio deve retornar nullptr sem criar entrada");
    exigir(c.inserir(IdSensor{"LT-101"}, {12, "%"}),
           "01: primeira insercao deve retornar true; complete inserir");
    exigir(!c.inserir(IdSensor{"LT-101"}, {99, "%"}), "01: recusar outra instancia da mesma tag");
    const auto* m = c.buscar(IdSensor{"LT-101"});
    exigir(m != nullptr, "01: buscar pela tag equivalente deve encontrar; complete buscar");
    exigir(m->valor == 12 && m->unidade == "%", "01: duplicata deve preservar valor E unidade originais");
    exigir(!c.inserir(IdSensor{"LT-101"}, {800, "cm"}), "01: mudar unidade nao cria nova chave");
    exigir(c.buscar(IdSensor{"LT-101"})->unidade == "%", "01: unidade original preservada");
    exigir(c.inserir(IdSensor{"LT-102"}, {0, "%"}), "01: aceitar segunda chave com leitura zero");
    exigir(c.inserir(IdSensor{"lt-101"}, {7, "%"}), "01: tag minuscula e distinta");
    exigir(c.quantidade() == 3 && c.ids().size() == 3, "01: contar chaves, nao tentativas");
    exigir(c.buscar(IdSensor{"LT-102"})->valor == 0, "01: zero nao representa ausencia");
    exigir(c.buscar(IdSensor{"outra"}) == nullptr && c.quantidade() == 3,
           "01: busca ausente nao pode alterar quantidade");
    exigir(c.ids().count(IdSensor{"LT-101"}) == 1, "01: conjunto de ids acompanha o mapa");
    Catalogo<Medicao> independente;
    exigir(independente.quantidade() == 0, "01: catalogos nao compartilham armazenamento");
    Catalogo<std::string> nomes;
    exigir(nomes.inserir(IdSensor{"X"}, "Bancada A"), "01: genericidade deve aceitar texto");
    const auto* nome = nomes.buscar(IdSensor{"X"});
    exigir(nome && *nome == "Bancada A", "01: busca generica preserva texto");
}
void ultimas(const Historico& h, std::size_t n, const std::vector<double>& valores) {
    const auto dados = h.ultimas(n);
    exigir(dados.size() == valores.size(), "02: ultimas(" + std::to_string(n) + ") retornou quantidade incorreta");
    for (std::size_t i = 0; i < dados.size(); ++i) {
        exigir(dados[i].valor == valores[i] && dados[i].unidade == "%",
               "02: ultimas deve preservar ordem de chegada, repeticoes e unidade");
    }
}
void extensao() {
    Catalogo<Medicao> c;
    exigir(!c.remover(IdSensor{"vazio"}), "02: remover no vazio deve retornar false");
    c.inserir(IdSensor{"A"}, {12, "%"}); c.inserir(IdSensor{"B"}, {20, "%"});
    exigir(!c.remover(IdSensor{"X"}) && c.quantidade() == 2, "02: remover ausente preserva as demais entradas");
    exigir(c.remover(IdSensor{"A"}), "02: remover chave existente deve retornar true; complete remover");
    exigir(c.quantidade() == 1 && c.buscar(IdSensor{"A"}) == nullptr, "02: remover deve alterar quantidade e busca");
    exigir(c.ids().count(IdSensor{"A"}) == 0 && c.ids().count(IdSensor{"B"}) == 1, "02: ids deve refletir a remocao");
    exigir(c.buscar(IdSensor{"B"}) && c.buscar(IdSensor{"B"})->valor == 20, "02: preservar outra chave");
    exigir(!c.remover(IdSensor{"A"}), "02: segunda remocao retorna false");
    exigir(c.inserir(IdSensor{"A"}, {33, "%"}), "02: permitir reinsercao depois de remover");
    exigir(c.buscar(IdSensor{"A"})->valor == 33, "02: reinsercao armazena novo valor");
    Historico h;
    ultimas(h, 0, {}); ultimas(h, 4, {});
    h.registrar({12, "%"}); h.registrar({12, "%"}); h.registrar({15, "%"});
    ultimas(h, 0, {}); ultimas(h, 1, {15}); ultimas(h, 2, {12, 15});
    ultimas(h, 3, {12, 12, 15}); ultimas(h, 100, {12, 12, 15});
    exigir(h.quantidade() == 3, "02: consultar ultimas nao pode descartar o historico");
    auto copia = h.ultimas(3);
    copia[0].valor = 999; copia.clear();
    ultimas(h, 3, {12, 12, 15});
    Medicao externa{40, "%"}; h.registrar(externa); externa.valor = 99;
    ultimas(h, 1, {40});
    exigir(c.remover(IdSensor{"A"}) && h.quantidade() == 4,
           "02: remover cadastro nao apaga ocorrencias do historico independente");
    h.registrar({5, "%"}); ultimas(h, 2, {40, 5});
}
int main(int argc, char** argv) {
    try {
        exigir(argc == 2, "informe etapa");
        const std::string etapa = argv[1];
        exigir(etapa == "00" || etapa == "01" || etapa == "02", "etapa invalida");
        base();
        if (etapa >= "01") catalogo();
        if (etapa == "02") extensao();
        std::cout << "OK C++ etapa " << etapa << '\n';
    } catch (const std::exception& e) {
        std::cerr << "FALHA C++: " << e.what() << '\n';
        return 1;
    }
}
